main.go

```diff
@@ -36,12 +51,51 @@ type Firehose struct {
 	processSettings ProcessSettings
 	storageBackend  StorageBackend
 	hedwigConsumer  *hedwig.QueueConsumer
+	messageBuffer []*hedwig.Message
+	flushLock sync.Mutex
+	flushCh chan struct{}
+}
+
+func (fp *Firehose) WriteMessage(wg *sync.WaitGroup, message *hedwig.Message) error {
+	defer wg.Done()
+	// TODO: serialization for hedwig.Message to []byte
+	var data []byte
+	// TODO: how to get store appenv? should file name end with time?
+	currentTime := time.Now()
+	uploadLocation := fmt.Sprintf("%s/%s/%s/%s", "store-appenv?", message.Type, currentTime.Format("2006/1/2"), "filetime?")
+	return fp.storageBackend.UploadFile(context.Background(), data, fp.processSettings.StagingBucket, uploadLocation)
+}
+
+func (fp *Firehose) flushCron() {
+	<-time.After(time.Duration(fp.processSettings.FlushAfter) * time.Second)
+	fp.flushLock.Lock()
+	defer fp.flushLock.Unlock()
+	var wg sync.WaitGroup
+	for _, msg := range fp.messageBuffer {
+		wg.Add(1)
+		fp.WriteMessage(&wg, msg)
+	}
+	// close channel to unblock all handleMessage functions (should ack msg from queue)
+	close(fp.flushCh)
+	fp.flushCh = make(chan struct{})
 }
 
-func (fp *Firehose) WriteMessages() {
+func (fp *Firehose) RunFollower(ctx context.Context) {
+	// run an infinite loop until canceled
+	// and call handleMessage
+	go fp.flushCron()
+	// should this be configureable?
+	lr := hedwig.ListenRequest{1, defaultVisibilityTimeoutS, 1}
+	fp.hedwigConsumer.ListenForMessages(ctx, lr)
 }
 
-func (fp *Firehose) RunFollower() {
+func (fp *Firehose) handleMessage(message *hedwig.Message) error {
```

Comment:
will this be the handler in the callbackregistry for all messages?

Reply:
yep